File: griffon-app/models/ru/dkolmogortsev/ControlPanelModel.java
package ru.dkolmogortsev;

import griffon.core.artifact.GriffonModel;
import griffon.metadata.ArtifactProviderFor;
import java.time.Duration;
import java.util.List;
import javafx.beans.property.ListProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.property.SimpleListProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.collections.FXCollections;
import org.codehaus.griffon.runtime.core.artifact.AbstractGriffonModel;
import org.reactfx.util.FxTimer;
import org.reactfx.util.Timer;
import ru.dkolmogortsev.utils.ElapsedTimeFormatter;

@ArtifactProviderFor(GriffonModel.class)
public class ControlPanelModel extends AbstractGriffonModel
{
    private SimpleIntegerProperty elapsedProperty = new SimpleIntegerProperty(0);
    private SimpleStringProperty currentTimeEntryId = new SimpleStringProperty();
    private SimpleBooleanProperty taskStarted = new SimpleBooleanProperty(false);
    private SimpleStringProperty timerTextProp = new SimpleStringProperty(
            ElapsedTimeFormatter.formatElapsed(elapsedProperty.get()));
    private SimpleStringProperty taskNameProperty = new SimpleStringProperty("Name");
    private SimpleStringProperty taskDescriptionProperty = new SimpleStringProperty("Description");
    private ListProperty<String> tasks = new SimpleListProperty<>();
    private Timer timer;
    public ControlPanelModel()
    {
        elapsedProperty.addListener((observable, oldValue, newValue) ->
        {
            System.out.println("changed");
            timerTextProp.set(ElapsedTimeFormatter.formatElapsed(newValue.intValue()));
        });
    }

    public SimpleStringProperty currentTimeEntryIdProperty()
    {
        return currentTimeEntryId;
    }

    public int getElapsedProperty()
    {
        return elapsedProperty.get();
    }

    public SimpleBooleanProperty taskStartedProperty()
    {
        return taskStarted;
    }

    public SimpleStringProperty taskNameProperty()
    {
        return taskNameProperty;
    }

    public SimpleStringProperty taskDescriptionProperty()
    {
        return taskDescriptionProperty;
    }

    public boolean isTaskStarted()
    {
        return taskStarted.get();
    }

    public SimpleStringProperty getTimerTextProp()
    {
        return timerTextProp;
    }

    public ListProperty<String> tasksProperty()
    {
        return tasks;
    }

    public void setList(List<String> tasks)
    {
        this.tasks.setValue(FXCollections.observableArrayList(tasks));
    }

    public void startTimer()
    {
        timer = FxTimer.runPeriodically(Duration.ofSeconds(1L), () -> elapsedProperty.set(elapsedProperty.get() + 1));
        taskStarted.set(true);
    }

    public void stopTimer()
    {
        elapsedProperty.set(0);
        timer.stop();
        taskStarted.set(false);
    }

}

File: griffon-app/models/ru/dkolmogortsev/ControlPanelModel.kt
package ru.dkolmogortsev

import griffon.core.artifact.GriffonModel
import griffon.metadata.ArtifactProviderFor
import javafx.beans.property.*
import javafx.collections.FXCollections
import org.codehaus.griffon.runtime.core.artifact.AbstractGriffonModel
import org.reactfx.util.FxTimer
import org.reactfx.util.Timer
import ru.dkolmogortsev.task.Task
import ru.dkolmogortsev.utils.formatToElapsed
import java.time.Duration

@ArtifactProviderFor(GriffonModel::class)
class ControlPanelModel : AbstractGriffonModel()
{
    private val elapsedProperty = SimpleLongProperty(0)
    private val currentTimeEntryIdProperty = SimpleLongProperty()
    private val taskStarted = SimpleBooleanProperty(false)
    val timerTextProp = SimpleStringProperty(elapsedProperty.get().formatToElapsed())
    private val taskNameProperty = SimpleStringProperty("Name")
    private val taskDescriptionProperty = SimpleStringProperty("Description")
    private val searchResultProperty = SimpleListProperty<Task>()
    var taskSelectedProperty: SimpleBooleanProperty = SimpleBooleanProperty(false)
    private lateinit var timer: Timer
    var timeEntryId
        get() = currentTimeEntryIdProperty.get()
        set(value) = currentTimeEntryIdProperty.set(value)

    init
    {
        elapsedProperty.addListener { _, _, newValue -> timerTextProp.set(newValue.toLong().formatToElapsed()) }
    }

    fun getElapsedProperty(): Long
    {
        return elapsedProperty.get()
    }

    fun taskStartedProperty(): SimpleBooleanProperty
    {
        return taskStarted
    }

    fun taskNameProperty(): SimpleStringProperty
    {
        return taskNameProperty
    }

    fun taskDescriptionProperty(): SimpleStringProperty
    {
        return taskDescriptionProperty
    }

    val isTaskStarted: Boolean
        get() = taskStarted.get()

    fun searchResultsProperty(): ListProperty<Task>
    {
        return searchResultProperty
    }

    fun setSearchResults(tasks: List<Task>)
    {
        this.searchResultProperty.value = FXCollections.observableArrayList(tasks)
    }

    fun startTimer()
    {
        timer = FxTimer
                .runPeriodically(Duration.ofSeconds(1L)) { elapsedProperty.value = getElapsedProperty() + 1 }
        taskStarted.set(true)
    }

    fun stopTimer()
    {
        elapsedProperty.set(0)
        timer.stop()
        taskStarted.set(false)
    }
}